tensorflow/core/common_runtime/direct_session.cc

```diff
@@ -542,8 +542,11 @@ Status DirectSession::RunInternal(
                             executor_step_count, &debugger_state));
   }
 
-  if (run_metadata != nullptr &&
-      options_.config.experimental().has_session_metadata()) {
+  if (run_metadata == nullptr) {
+    return absl::InternalError("Metadata output is not set.");
+  }
```

Comment:
Unfortunately, this is a breaking change to the API, which results in at least some of the test failures. The `run_metadata` has to be optional. 

But the fundamental observation in the PR is correct: it is possible to trigger a segfault or other undefined behavior via particular combinations of `RunOptions` and a null `run_metadata`.

Unfortunately that means that we need guards on each subsequent use of `run_metadata`. Would you be able to take this on? Thanks!

Reply:
Thank you for the revision! I'll fix it in near future.